File: Ontology/ClassMembershipEvaluator.cs
using CommonDomainObjects;
using System;
using System.Collections.Generic;
using System.Linq;

namespace Ontology
{
    public class ClassMembershipEvaluator: IClassMembershipEvaluator
    {
        private readonly IOntology                                           _ontology;
        private readonly IDictionary<IClass, IClassExpression>               _classDefinitions;
        private readonly IList<IClass>                                       _definedClasses;
        private readonly IDictionary<string, IClass>                         _classes;
        private readonly ILookup<INamedIndividual, IClassExpression>         _classAssertions;
        private readonly ILookup<INamedIndividual, IObjectPropertyAssertion> _objectPropertyAssertions;
        private readonly ILookup<INamedIndividual, IDataPropertyAssertion>   _dataPropertyAssertions;
        private readonly ILookup<IClassExpression, IClassExpression>         _superClassExpressions;
        private readonly ILookup<IClassExpression, IClassExpression>         _subClassExpressions;
        private readonly ILookup<IClassExpression, IClassExpression>         _disjointClassExpressions;
        private readonly IDictionary<object, ISet<IClass>>                   _classifications;

        private class ClassVisitor: ClassExpressionVisitor
        {
            private readonly Action<IClass> _action;

            public ClassVisitor(
                Action<IClass> action
                )
            {
                _action = action;
            }

            public override void Visit(
                IClass @class
                ) => _action(@class);
        }

        private struct ClassComparer: IComparer<IClass>
        {
            private readonly IDictionary<IClass, int> _longestPaths;

            public ClassComparer(
                IDictionary<IClass, int> longestPaths
                )
            {
                _longestPaths = longestPaths;
            }

            int IComparer<IClass>.Compare(
                IClass x,
                IClass y) => _longestPaths[x].CompareTo(_longestPaths[y]);
        }

        public ClassMembershipEvaluator(
            IOntology                         ontology,
            IDictionary<object, ISet<IClass>> classifications
            )
        {
            _ontology                 = ontology;
            _classifications          = classifications;
            _classes = _ontology.Get<IClass>().ToDictionary(
                @class => @class.Iri);
            _classAssertions          = _ontology.Get<IClassAssertion>().ToLookup(
                classAssertion => classAssertion.NamedIndividual,
                classAssertion => classAssertion.ClassExpression);
            _objectPropertyAssertions = _ontology.Get<IObjectPropertyAssertion>().ToLookup(
                objectPropertyAssertion => objectPropertyAssertion.SourceIndividual);
            _dataPropertyAssertions   = _ontology.Get<IDataPropertyAssertion>().ToLookup(
                dataPropertyAssertion => dataPropertyAssertion.SourceIndividual);
            _superClassExpressions    = _ontology.Get<ISubClassOf>().ToLookup(
                subClassOf => subClassOf.SubClassExpression,
                subClassOf => subClassOf.SuperClassExpression);
            _subClassExpressions      = _ontology.Get<ISubClassOf>().ToLookup(
                subClassOf => subClassOf.SuperClassExpression,
                subClassOf => subClassOf.SubClassExpression);
            _classDefinitions         = (
                from @class in _ontology.Get<IClass>()
                from equivalentClasses in _ontology.Get<IEquivalentClasses>()
                where equivalentClasses.ClassExpressions.Contains(@class)
                from classExpression in equivalentClasses.ClassExpressions
                where !(classExpression is IClass)
                group classExpression by @class into classExpressionsGroupedbyClass
                select
                (
                    Class          : classExpressionsGroupedbyClass.Key,
                    ClassExpression: classExpressionsGroupedbyClass.First()
                )).ToDictionary(
                    definition => definition.Class,
                    definition => definition.ClassExpression);

            var disjointPairs = (
                from disjointClasses in _ontology.Get<IDisjointClasses>()
                from classExpression1 in disjointClasses.ClassExpressions
                from classExpression2 in disjointClasses.ClassExpressions
                where classExpression1 != classExpression2
                select
                (
                    ClassExpression1: classExpression1,
                    ClassExpression2: classExpression2
                )).ToList();

            for(var index = 0;index < disjointPairs.Count;++index)
            {
                var (classExpression1, classExpression2) = disjointPairs[index];
                _subClassExpressions[classExpression2].ForEach(
                    subclassExpression => disjointPairs.Add(
                        (
                            ClassExpression1: classExpression1,
                            ClassExpression2: subclassExpression
                        )));
            }

            _disjointClassExpressions = disjointPairs.ToLookup(
                disjointPair => disjointPair.ClassExpression1,
                disjointPair => disjointPair.ClassExpression2);

            HashSet<IClass> adjacent = null;
            var empty = new HashSet<IClass>();
            var adjacencyList = _ontology.Get<IClass>()
                .ToDictionary(
                    @class => @class,
                    @class => empty);

            var classVisitor = new ClassExpressionNavigator(new ClassVisitor(@class => adjacent.Add(@class)));

            _classDefinitions.ForEach(
                pair =>
                {
                    adjacent = new HashSet<IClass>();
                    pair.Value.Accept(classVisitor);
                    adjacencyList[pair.Key] = adjacent;
                });

            _definedClasses = (
                from @class in adjacencyList.TopologicalSort()
                join definition in _classDefinitions on @class equals definition.Key
                select definition.Key
                ).ToList();
        }

        bool IClassMembershipEvaluator.Evaluate(
            IClass @class,
            object individual
            ) => Classify(individual).Contains(@class);

        bool IClassMembershipEvaluator.Evaluate(
            IObjectIntersectionOf objectIntersectionOf,
            object                individual
            ) => objectIntersectionOf.ClassExpressions.All(classExpression => classExpression.Evaluate(
                this,
                individual));

        bool IClassMembershipEvaluator.Evaluate(
            IObjectUnionOf objectUnionOf,
            object         individual
            ) => objectUnionOf.ClassExpressions.Any(classExpression => classExpression.Evaluate(
                this,
                individual));

        bool IClassMembershipEvaluator.Evaluate(
            IObjectComplementOf objectComplementOf,
            object              individual
            ) => !objectComplementOf.ClassExpression.Evaluate(
                this,
                individual);

        bool IClassMembershipEvaluator.Evaluate(
            IObjectOneOf objectOneOf,
            object       individual
            ) => objectOneOf.Individuals.Any(
                member => AreEqual(
                    individual,
                    member));

        bool IClassMembershipEvaluator.Evaluate(
            IObjectSomeValuesFrom objectSomeValuesFrom,
            object                individual
            ) => ObjectPropertyValues(
                objectSomeValuesFrom.ObjectPropertyExpression,
                individual).Any(
                    value => objectSomeValuesFrom.ClassExpression.Evaluate(
                        this,
                        value));

        bool IClassMembershipEvaluator.Evaluate(
            IObjectAllValuesFrom objectAllValuesFrom,
            object               individual
            ) => ObjectPropertyValues(
                objectAllValuesFrom.ObjectPropertyExpression,
                individual).All(
                    value => objectAllValuesFrom.ClassExpression.Evaluate(
                        this,
                        value));

        bool IClassMembershipEvaluator.Evaluate(
            IObjectHasValue objectHasValue,
            object          individual
            ) => ObjectPropertyValues(
                objectHasValue.ObjectPropertyExpression,
                individual).Any(
                    value => AreEqual(
                        objectHasValue.Individual,
                        value));

        bool IClassMembershipEvaluator.Evaluate(
            IObjectHasSelf objectHasValue,
            object         individual
            ) => ObjectPropertyValues(
                objectHasValue.ObjectPropertyExpression,
                individual).Any(
                    value => AreEqual(
                        individual,
                        value));

        bool IClassMembershipEvaluator.Evaluate(
            IObjectMinCardinality objectMinCardinality,
            object                individual
            ) => ObjectPropertyValues(
                objectMinCardinality.ObjectPropertyExpression,
                individual).Count(
                    value => (objectMinCardinality.ClassExpression ?? ReservedVocabulary.Thing).Evaluate(
                        this,
                        value)) >= objectMinCardinality.Cardinality;

        bool IClassMembershipEvaluator.Evaluate(
            IObjectMaxCardinality objectMaxCardinality,
            object                individual
            ) => ObjectPropertyValues(
                objectMaxCardinality.ObjectPropertyExpression,
                individual).Count(
                    value => (objectMaxCardinality.ClassExpression ?? ReservedVocabulary.Thing).Evaluate(
                        this,
                        value)) <= objectMaxCardinality.Cardinality;

        bool IClassMembershipEvaluator.Evaluate(
            IObjectExactCardinality objectExactCardinality,
            object                  individual
            ) => ObjectPropertyValues(
                objectExactCardinality.ObjectPropertyExpression,
                individual).Count(
                    value => (objectExactCardinality.ClassExpression ?? ReservedVocabulary.Thing).Evaluate(
                        this,
                        value)) == objectExactCardinality.Cardinality;

        bool IClassMembershipEvaluator.Evaluate(
            IDataSomeValuesFrom dataSomeValuesFrom,
            object              individual
            ) => DataPropertyValues(
                dataSomeValuesFrom.DataPropertyExpression,
                individual).Any(dataSomeValuesFrom.DataRange.HasMember);

        bool IClassMembershipEvaluator.Evaluate(
            IDataAllValuesFrom dataAllValuesFrom,
            object             individual
            ) => DataPropertyValues(
                dataAllValuesFrom.DataPropertyExpression,
                individual).All(dataAllValuesFrom.DataRange.HasMember);

        bool IClassMembershipEvaluator.Evaluate(
            IDataHasValue dataHasValue,
            object        individual
            ) => DataPropertyValues(
                dataHasValue.DataPropertyExpression,
                individual).Contains(dataHasValue.Value);

        bool IClassMembershipEvaluator.Evaluate(
            IDataMinCardinality dataMinCardinality,
            object              individual
            ) => DataPropertyValues(
                dataMinCardinality.DataPropertyExpression,
                individual).Count(value => dataMinCardinality.DataRange?.HasMember(value) ?? true) >= dataMinCardinality.Cardinality;

        bool IClassMembershipEvaluator.Evaluate(
            IDataMaxCardinality dataMaxCardinality,
            object              individual
            ) => DataPropertyValues(
                dataMaxCardinality.DataPropertyExpression,
                individual).Count(value => dataMaxCardinality.DataRange?.HasMember(value) ?? true) <= dataMaxCardinality.Cardinality;

        bool IClassMembershipEvaluator.Evaluate(
            IDataExactCardinality dataExactCardinality,
            object                individual
            ) => DataPropertyValues(
                dataExactCardinality.DataPropertyExpression,
                individual).Count(value => dataExactCardinality.DataRange?.HasMember(value) ?? true) == dataExactCardinality.Cardinality;

        private bool AreEqual(
            object lhs,
            object rhs
            )
        {
            var commonClassExpressions = Classify(lhs);
            commonClassExpressions.IntersectWith(Classify(rhs));
            var hasKeys =
                from commonClassExpression in commonClassExpressions
                join hasKey in _ontology.Get<IHasKey>() on commonClassExpression equals hasKey.ClassExpression
                select hasKey;
            return
                hasKeys.Any() &&
                hasKeys.All(
                    hasKey => hasKey.DataPropertyExpressions.All(
                        dataPropertyExpression => AreEqual(
                            dataPropertyExpression,
                            lhs,
                            rhs)));
        }

        private bool AreEqual(
            IDataPropertyExpression dataPropertyExpression,
            object lhs,
            object rhs
            )
        {
            var lhsValues = DataPropertyValues(
                dataPropertyExpression,
                lhs);
            var rhsValues = DataPropertyValues(
                dataPropertyExpression,
                rhs);
            return dataPropertyExpression is IFunctionalDataProperty ?
                Equals(
                    lhsValues.FirstOrDefault(),
                    rhsValues.FirstOrDefault()) :
                lhsValues.ToHashSet().SetEquals(lhsValues.ToHashSet());
        }

        public ISet<IClass> Classify(
            object individual
            )
        {
            if(_classifications.TryGetValue(
                individual,
                out ISet<IClass> classes))
                return classes;

            _classifications[individual] = classes = new HashSet<IClass>();
            var candidates = _definedClasses.ToHashSet();

            switch(individual)
            {
                case INamedIndividual namedIndividual:
                    _classAssertions[namedIndividual].ForEach(classExpression => Classify(
                        classes,
                        candidates,
                        individual,
                        classExpression));
                    break;
                case IIndividual iindividual:
                    if(_classes.TryGetValue(
                        iindividual.ClassIri,
                        out IClass @class))
                        Classify(
                            classes,
                            candidates,
                            individual,
                            @class);
                    break;
                default:
                    (
                        from @class1 in _classes.Values
                        join type in individual.GetTypes() on @class1.Iri equals type.FullName
                        select @class1
                    ).ForEach(@class1 => Classify(
                        classes,
                        candidates,
                        individual,
                        @class1));
                    break;
            }

            foreach(var definedClass in _definedClasses)
                if(candidates.Contains(definedClass) &&
                   _classDefinitions[definedClass].Evaluate(
                    this,
                    individual))
                    Classify(
                        classes,
                        candidates,
                        individual,
                        definedClass);

            return classes;
        }

        private void Classify(
            ISet<IClass>     classes,
            ISet<IClass>     candidates,
            object           individual,
            IClassExpression classExpression
            )
        {
            if(classExpression is IClass @class)
            {
                if(!classes.Add(@class))
                    // Class Expression already processed.
                    return;

                // Prune candidates.
                candidates.Remove(@class);
                candidates.ExceptWith(_disjointClassExpressions[classExpression].OfType<IClass>());

                _superClassExpressions[classExpression].ForEach(superClassExpression => Classify(
                    classes,
                    candidates,
                    individual,
                    superClassExpression));
            }
            else if(classExpression is IObjectIntersectionOf objectIntersectionOf)
                objectIntersectionOf.ClassExpressions
                    .ForEach(componentClassExpression => Classify(
                        classes,
                        candidates,
                        individual,
                        componentClassExpression));
        }

        public IEnumerable<object> ObjectPropertyValues(
            IObjectPropertyExpression objectPropertyExpression,
            object                    individual
            ) => individual is INamedIndividual namedIndividual ?
                _objectPropertyAssertions[namedIndividual]
                    .Where(objectPropertyAssertion => objectPropertyAssertion.ObjectPropertyExpression == objectPropertyExpression)
                    .Select(objectPropertyAssertion => objectPropertyAssertion.TargetIndividual) :
                objectPropertyExpression.Values(individual);

        public IEnumerable<object> DataPropertyValues(
            IDataPropertyExpression dataPropertyExpression,
            object individual
            ) => individual is INamedIndividual namedIndividual ?
                _dataPropertyAssertions[namedIndividual]
                    .Where(dataPropertyAssertion => dataPropertyAssertion.DataPropertyExpression == dataPropertyExpression)
                    .Select(dataPropertyAssertion => dataPropertyAssertion.TargetValue) :
                dataPropertyExpression.Values(individual);
    }
}

File: Web/ClientApp/src/app/Ontology/ClassMembershipEvaluator.ts
import { LongestPaths } from "./AdjacencyList";
import { ClassExpressionNavigator } from "./ClassExpressionNavigator";
import { Group } from "./Group";
import { IClass } from "./IClass";
import { IClassExpression } from "./IClassExpression";
import { ClassExpressionVisitor } from "./IClassExpressionVisitor";
import { IClassMembershipEvaluator } from "./IClassMembershipEvaluator";
import { IDataAllValuesFrom } from "./IDataAllValuesFrom";
import { IDataExactCardinality, IDataMaxCardinality, IDataMinCardinality } from "./IDataCardinality";
import { IDataHasValue } from "./IDataHasValue";
import { IDataSomeValuesFrom } from "./IDataSomeValuesFrom";
import { IHasKey } from "./IHasKey";
import { INamedIndividual } from "./INamedIndividual";
import { IObjectAllValuesFrom } from "./IObjectAllValuesFrom";
import { IObjectExactCardinality, IObjectMaxCardinality, IObjectMinCardinality } from "./IObjectCardinality";
import { IObjectComplementOf } from "./IObjectComplementOf";
import { IObjectHasSelf } from "./IObjectHasSelf";
import { IObjectHasValue } from "./IObjectHasValue";
import { IObjectIntersectionOf } from "./IObjectIntersectionOf";
import { IObjectOneOf } from "./IObjectOneOf";
import { IObjectSomeValuesFrom } from "./IObjectSomeValuesFrom";
import { IObjectUnionOf } from "./IObjectUnionOf";
import { IOntology } from "./IOntology";
import { IDataPropertyExpression } from "./IPropertyExpression";
import { IStore, Store, StoreDecorator } from "./IStore";
import { TransitiveClosure2 } from './TransitiveClosure';

class ClassVisitor extends ClassExpressionVisitor
{
    private readonly _action: (class$: IClass) => { };

    constructor(
        action: (class$: IClass) => {}
        )
    {
        super();
        this._action = action;
    }

    Visit(
        class$: IClass
        )
    {
        this._action(class$);
    }
}

export class ClassMembershipEvaluator implements IClassMembershipEvaluator
{
    public  readonly Store                     : IStore;
    private readonly _ontology                 : IOntology;
    private readonly _classes                  = new Map<string, IClass>();
    private readonly _classAssertions          : Map<INamedIndividual, IClassExpression[]>;
    private readonly _hasKeys                  : IHasKey[];
    private readonly _superClassExpressions    : Map<IClassExpression, IClassExpression[]>;
    private readonly _subClassExpressions      : Map<IClassExpression, IClassExpression[]>;
    private readonly _disjointClassExpressions : Map<IClassExpression, IClassExpression[]>;
    private readonly _functionalDataProperties = new Set<IDataPropertyExpression>();
    private readonly _classDefinitions         : Map<IClass, IClassExpression[]>;
    private readonly _definedClasses           : IClass[];
    private readonly _classifications          : Map<object, Set<IClass>>;

    constructor(
        ontology        : IOntology,
        classifications?: Map<object, Set<IClass>>
        )
    {
        this._ontology        = ontology;
        this._classifications = classifications;
        this.Store = new StoreDecorator(
            ontology,
            new Store());

        if(!this._classifications)
            this._classifications = new Map<object, Set<IClass>>();

        for(let class$ of ontology.Get(ontology.IsAxiom.IClass))
            this._classes.set(
                class$.Iri,
                class$);

        this._classAssertions = Group(
            ontology.Get(ontology.IsAxiom.IClassAssertion),
            classAssertion => classAssertion.NamedIndividual,
            classAssertion => classAssertion.ClassExpression);
        this._hasKeys = [...ontology.Get(ontology.IsAxiom.IHasKey)];
        this._superClassExpressions = Group(
            ontology.Get(ontology.IsAxiom.ISubClassOf),
            subClassOf => subClassOf.SubClassExpression,
            subClassOf => subClassOf.SuperClassExpression);
        this._subClassExpressions = Group(
            ontology.Get(ontology.IsAxiom.ISubClassOf),
            subClassOf => subClassOf.SuperClassExpression,
            subClassOf => subClassOf.SubClassExpression);

        let disjointPairs: [IClassExpression, IClassExpression][] = [];
        for(let disjointClasses of ontology.Get(ontology.IsAxiom.IDisjointClasses))
            for(let classExpression1 of disjointClasses.ClassExpressions)
                for(let classExpression2 of disjointClasses.ClassExpressions)
                    if(classExpression1 !== classExpression2)
                        disjointPairs.push(
                            [
                                classExpression1,
                                classExpression2
                            ]);

        for(let index = 0; index < disjointPairs.length; ++index)
        {
            let [classExpression1, classExpression2] = disjointPairs[index];
            let subClassExpressions = this._subClassExpressions.get(classExpression2);
            if(subClassExpressions)
                subClassExpressions.forEach(
                    subclassExpression => disjointPairs.push(
                        [
                            classExpression1,
                            subclassExpression
                        ]));
        }

        this._disjointClassExpressions = Group(
            disjointPairs,
            disjointPair => disjointPair[0],
            disjointPair => disjointPair[1]);

        for(let functionalDataProperty of ontology.Get(ontology.IsAxiom.IFunctionalDataProperty))
            this._functionalDataProperties.add(functionalDataProperty.DataPropertyExpression);

        let adjacencyMatrix = new Map<IClass, Map<IClass, boolean>>();
        let classes = [...this._ontology.Get(this._ontology.IsAxiom.IClass)];
        for(let rowClass of classes)
        {
            const row = new Map<IClass, boolean>();
            adjacencyMatrix.set(
                rowClass,
                row);

            for(let columnClass of classes)
                row.set(
                    columnClass,
                    rowClass === columnClass);
        }

        for(let equivalentClassExpressions of ontology.Get(ontology.IsAxiom.IEquivalentClasses))
        {
            let equivalentClasses = <IClass[]>equivalentClassExpressions.ClassExpressions.filter(classExpression => ontology.IsAxiom.IClass(classExpression));
            for(let class1 of equivalentClasses)
                for(let class2 of equivalentClasses)
                {
                    adjacencyMatrix.get(class1).set(
                        class2,
                        true);
                    adjacencyMatrix.get(class2).set(
                        class1,
                        true);
                }
        }

        let transitiveClosure = TransitiveClosure2(adjacencyMatrix);

        let definitions: [IClass, IClassExpression][] = [];
        for(let equivalentClasses of ontology.Get(ontology.IsAxiom.IEquivalentClasses))
            for(let class$ of equivalentClasses.ClassExpressions.filter(classExpression => ontology.IsAxiom.IClass(classExpression)))
            {
                for(let classExpression of equivalentClasses.ClassExpressions.filter(classExpression => !ontology.IsAxiom.IClass(classExpression)))
                    for(let entry of transitiveClosure.get(<IClass>class$))
                        if(entry[1])
                            definitions.push(
                                [
                                    entry[0],
                                    classExpression
                                ]);
                break;
            }
        this._classDefinitions = Group(
            definitions,
            definition => definition[0],
            definition => definition[1]);

        let adjacent: Set<IClass> = null;
        let empty = new Set<IClass>();
        let adjacencyList = new Map<IClass, Set<IClass>>();
        for(let class$ of ontology.Get(ontology.IsAxiom.IClass))
            adjacencyList.set(
                class$,
                empty);

        let classVisitor = new ClassExpressionNavigator(new ClassVisitor(class$ => adjacent.add(class$)));

        for(let [class$, classExpression] of this._classDefinitions)
        {
            adjacent = new Set<IClass>();
            classExpression[0].Accept(classVisitor);
            adjacencyList.set(
                class$,
                adjacent);
        }

        let longestPaths = LongestPaths(adjacencyList);
        this._definedClasses = [...this._classDefinitions.keys()]
            .sort((a, b) => longestPaths.get(b) - longestPaths.get(a));
    }

    Class(
        class$    : IClass,
        individual: object
        ): boolean
    {
        return this.Classify(individual).has(class$);
    }

    ObjectIntersectionOf(
        objectIntersectionOf: IObjectIntersectionOf,
        individual          : object
        ): boolean
    {
        return objectIntersectionOf.ClassExpressions.every(classExpression => classExpression.Evaluate(
            this,
            individual));
    }

    ObjectUnionOf(
        objectUnionOf: IObjectUnionOf,
        individual   : object
        ): boolean
    {
        return objectUnionOf.ClassExpressions.some(classExpression => classExpression.Evaluate(
            this,
            individual));
    }

    ObjectComplementOf(
        objectComplementOf: IObjectComplementOf,
        individual        : object
        ): boolean
    {
        return !objectComplementOf.ClassExpression.Evaluate(
            this,
            individual);
    }

    ObjectOneOf(
        objectOneOf: IObjectOneOf,
        individual : object
        ): boolean
    {
        return objectOneOf.Individuals.some(
            member => this.AreEqual(
                individual,
                member));
    }

    ObjectSomeValuesFrom(
        objectSomeValuesFrom: IObjectSomeValuesFrom,
        individual          : object
        ): boolean
    {
        return this.Store.ObjectPropertyValues(
            objectSomeValuesFrom.ObjectPropertyExpression,
            individual).some(
                value => objectSomeValuesFrom.ClassExpression.Evaluate(
                    this,
                    value));
    }

    ObjectAllValuesFrom(
        objectAllValuesFrom: IObjectAllValuesFrom,
        individual         : object
        ): boolean
    {
        return this.Store.ObjectPropertyValues(
            objectAllValuesFrom.ObjectPropertyExpression,
            individual).every(
                value => objectAllValuesFrom.ClassExpression.Evaluate(
                    this,
                    value));
    }

    ObjectHasValue(
        objectHasValue: IObjectHasValue,
        individual    : object
        ): boolean
    {
        return this.Store.ObjectPropertyValues(
            objectHasValue.ObjectPropertyExpression,
            individual).some(
                value => this.AreEqual(
                    objectHasValue.Individual,
                    value));
    }

    ObjectHasSelf(
        objectHasSelf: IObjectHasSelf,
        individual   : object
        ): boolean
    {
        return this.Store.ObjectPropertyValues(
            objectHasSelf.ObjectPropertyExpression,
            individual).some(
                value => this.AreEqual(
                    individual,
                    value));
    }

    ObjectMinCardinality(
        objectMinCardinality: IObjectMinCardinality,
        individual          : object
        ): boolean
    {
        return this.Store.ObjectPropertyValues(
            objectMinCardinality.ObjectPropertyExpression,
            individual).reduce(
                (count: number, value: object) =>
                    !objectMinCardinality.ClassExpression ||
                    objectMinCardinality.ClassExpression.Evaluate(
                        this,
                        value) ? count + 1 : count,
                0) >= objectMinCardinality.Cardinality;
    }

    ObjectMaxCardinality(
        objectMaxCardinality: IObjectMaxCardinality,
        individual          : object
        ): boolean
    {
        return this.Store.ObjectPropertyValues(
            objectMaxCardinality.ObjectPropertyExpression,
            individual).reduce(
                (count: number, value: object) =>
                    !objectMaxCardinality.ClassExpression ||
                    objectMaxCardinality.ClassExpression.Evaluate(
                        this,
                        value) ? count + 1 : count,
                0) <= objectMaxCardinality.Cardinality;
    }

    ObjectExactCardinality(
        objectExactCardinality: IObjectExactCardinality,
        individual            : object
        ): boolean
    {
        return this.Store.ObjectPropertyValues(
            objectExactCardinality.ObjectPropertyExpression,
            individual).reduce(
                (count: number, value: object) =>
                    !objectExactCardinality.ClassExpression ||
                    objectExactCardinality.ClassExpression.Evaluate(
                        this,
                        value) ? count + 1 : count,
                0) === objectExactCardinality.Cardinality;
    }

    DataSomeValuesFrom(
        dataSomeValuesFrom: IDataSomeValuesFrom,
        individual        : object
        ): boolean
    {
        return this.Store.DataPropertyValues(
            dataSomeValuesFrom.DataPropertyExpression,
            individual).some(value => dataSomeValuesFrom.DataRange.HasMember(value));
    }

    DataAllValuesFrom(
        dataAllValuesFrom: IDataAllValuesFrom,
        individual       : object
        ): boolean
    {
        return this.Store.DataPropertyValues(
            dataAllValuesFrom.DataPropertyExpression,
            individual).every(value => dataAllValuesFrom.DataRange.HasMember(value));
    }

    DataHasValue(
        dataHasValue: IDataHasValue,
        individual  : object
        ): boolean
    {
        return this.Store.DataPropertyValues(
            dataHasValue.DataPropertyExpression,
            individual).includes(dataHasValue.Value);
    }

    DataMinCardinality(
        dataMinCardinality: IDataMinCardinality,
        individual        : object
        ): boolean
    {
        return this.Store.DataPropertyValues(
            dataMinCardinality.DataPropertyExpression,
            individual).reduce(
                (count: number, value: object) =>
                    !dataMinCardinality.DataRange ||
                    dataMinCardinality.DataRange.HasMember(value) ? count + 1 : count,
                0) >= dataMinCardinality.Cardinality;
    }

    DataMaxCardinality(
        dataMaxCardinality: IDataMaxCardinality,
        individual        : object
        ): boolean
    {
        return this.Store.DataPropertyValues(
            dataMaxCardinality.DataPropertyExpression,
            individual).reduce(
                (count: number, value: object) =>
                    !dataMaxCardinality.DataRange ||
                    dataMaxCardinality.DataRange.HasMember(value) ? count + 1 : count,
                0) <= dataMaxCardinality.Cardinality;
    }

    DataExactCardinality(
        dataExactCardinality: IDataExactCardinality,
        individual          : object
        ): boolean
    {
        return this.Store.DataPropertyValues(
            dataExactCardinality.DataPropertyExpression,
            individual).reduce(
                (count: number, value: object) =>
                    !dataExactCardinality.DataRange ||
                    dataExactCardinality.DataRange.HasMember(value) ? count + 1 : count,
                0) === dataExactCardinality.Cardinality;
    }

    Classify(
        individual: object
        ): Set<IClass>
    {
        let classes = this._classifications.get(individual);
        if(classes)
            return classes;

        classes = new Set<IClass>();
        this._classifications.set(
            individual,
            classes);

        let candidates = new Set<IClass>(this._definedClasses);

        if(this._ontology.IsAxiom.INamedIndividual(individual))
        {
            let assertedClassExpressions = this._classAssertions.get(individual);
            if(assertedClassExpressions)
                assertedClassExpressions
                    .forEach(assertedClassExpression => this.ApplyClassExpression(
                        classes,
                        candidates,
                        individual,
                        assertedClassExpression));
        }
        else if('ClassIri' in individual)
        {
            let class$ = this._classes.get(individual['ClassIri']);
            if(class$)
                this.ApplyClassExpression(
                    classes,
                    candidates,
                    individual,
                    class$);
        }

        for(let definedClass of this._definedClasses)
            if(candidates.has(definedClass) &&
                this._classDefinitions.get(definedClass)[0].Evaluate(
                    this,
                    individual))
                this.ApplyClassExpression(
                    classes,
                    candidates,
                    individual,
                    definedClass);

        return classes;
    }

    private ApplyClassExpression(
        classes         : Set<IClass>,
        candidates      : Set<IClass>,
        individual      : object,
        classExpression : IClassExpression
        ): void
    {
        if(this._ontology.IsAxiom.IClass(classExpression))
        {
            if(classes.has(classExpression))
                // Class Expression already processed.
                return;

            classes.add(classExpression);

            // Prune candidates.
            candidates.delete(classExpression);

            let disjointClassExpressions = this._disjointClassExpressions.get(classExpression);
            if(disjointClassExpressions)
                for(let disjointClassExpression of disjointClassExpressions)
                    if(this._ontology.IsAxiom.IClass(disjointClassExpression))
                        candidates.delete(disjointClassExpression);;


            let superClassExpressions = this._superClassExpressions.get(classExpression);
            if(superClassExpressions)
                superClassExpressions.forEach(superClassExpression => this.ApplyClassExpression(
                    classes,
                    candidates,
                    individual,
                    superClassExpression));
        }

        if(this._ontology.IsClassExpression.IObjectIntersectionOf(classExpression))
            classExpression.ClassExpressions
                .forEach(componentClassExpression => this.ApplyClassExpression(
                    classes,
                    candidates,
                    individual,
                    componentClassExpression));
    }

    public AreEqual(
        lhs: object,
        rhs: object
        ): boolean
    {
        const hasKeys = this._hasKeys
            .filter(hasKey => hasKey.ClassExpression.Evaluate(this, lhs) && hasKey.ClassExpression.Evaluate(this, rhs));

        return hasKeys.length &&
            hasKeys.every(
                hasKey => hasKey.DataPropertyExpressions.every(
                    dataPropertyExpression => this.DataPropertyExpressionAreEqual(
                        dataPropertyExpression,
                        lhs,
                        rhs)));
    }

    private DataPropertyExpressionAreEqual(
        dataPropertyExpression: IDataPropertyExpression,
        lhs: object,
        rhs: object
        ): boolean
    {
        if(this._functionalDataProperties.has(dataPropertyExpression))
            return this.Store.DataPropertyValue(
                dataPropertyExpression,
                lhs) === this.Store.DataPropertyValue(
                    dataPropertyExpression,
                    rhs);
        else
        {
            let lhsValues = new Set<object>(this.Store.DataPropertyValues(
                dataPropertyExpression,
                lhs));
            let rhsValues = new Set<object>(this.Store.DataPropertyValues(
                dataPropertyExpression,
                rhs));

            return lhsValues.size === rhsValues.size &&
                [...lhsValues].every(lhsValue => rhsValues.has(lhsValue));
        }
    }
}
